component/wrapper.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2016-2025, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package component
+
+import (
+	"fmt"
+
+	p "github.com/pulumi/pulumi-go-provider"
+	"github.com/pulumi/pulumi-go-provider/infer"
+	"github.com/pulumi/pulumi-go-provider/middleware/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+)
+
+// Resource is a type alias for the inferred component resource type.
+type Resource = infer.InferredComponent
+
+// ComponentFn is a type alias for infer.ComponentFn which represents
+// a function that creates a component resource.
+type ComponentFn[I any, O pulumi.ComponentResource] = infer.ComponentFn[I, O]
+
+// ProgramComponent allows us to create a component resource that can be understood by the
+// underlying provider. This is a wrapper around the infer.ProgramComponent function.
+func ProgramComponent[I any, O pulumi.ComponentResource](fn ComponentFn[I, O]) Resource {
```

Comment:
Why both ComponentFn and ProgramComponent? Why not combine them into

```go
// ProgramComponent allows us to create a component resource that can be understood by the
// underlying provider. This is a wrapper around the infer.ProgramComponent function.
func ProgramComponent[I any, O pulumi.ComponentResource](
	fn func(*pulumi.Context, string, I, ...pulumi.ResourceOption) (O, error)) Resource {
	return infer.ProgramComponent[I, O](fn)
}
```

and then usage is

```go
err := component.ProviderHost(
		component.WithName("go-components"),
		component.WithResources(
			component.ProgramComponent[RandomComponentArgs, *RandomComponent](NewRandomComponent)),
	)
```

Reply:
Ah, wait, it probably the same thing, ComponentFn is just an alias.